_posts/2025-05-05-Exposing-WASM-binaries-as-MCP-tools.adoc

```diff
@@ -0,0 +1,125 @@
+
+== Benefits of WASM-based MCP tools
+
+* Near-native performance for compute-intensive operations
+* Ability to reuse existing tools written in languages like C, C++, or Rust
+* Secure execution within WildFly's sandbox environment
+* Cross-platform compatibility without recompilation
+
+== Defining WASM Binaries for MCP Tool Integration
+
+Before a WASM binary can be exposed as an MCP tool, it needs to be defined within WildFly's WASM subsystem. This subsystem acts as a registry for your WASM modules.
+It is based on the https://github.com/extism/chicory-sdk[chicory-sdk], an experimental Java SDK using the https://github.com/dylibso/chicory[JVM-native Chicory] runtime.
+
+=== Configuration Methods
+
+You can configure these binaries using either the WildFly CLI or by directly editing the XML configuration (`standalone.xml`).
+
+==== Using CLI
+
+Add a WASM binary definition via the CLI like this:
+
+[source,bash]
+----
+/subsystem=wasm/wasm-tool=my-wasm-tool:add(path="/path/to/your/binary.wasm")
+----
+
+*   `my-wasm-tool`: A unique name for this binary definition.
+*   `path`: The location of the `.wasm` file. This path is relative to the `relative-to` attribute if specified.
+
+==== Using XML Configuration
+
+Alternatively, add the definition within the `<subsystem xmlns="urn:wildfly:wasm:1.0">` element in your configuration file:
+
+[source,xml]
+----
+<subsystem xmlns="urn:jboss:domain:wasm:1.0">
+    <wasm-tool name="my-wasm-tool" path="path/relative/to/module/root/binary.wasm"/>
+    <!-- Other binary definitions -->
+</subsystem>
+----
+
+Ensure the `path` is correctly specified relative to the root of the deployment module.
+
+== Next Steps: CDI Integration
+
+Once defined in the WASM subsystem, these binaries can be injected using CDI.
+The first way you can use them is to get them by name as an `org.wildfly.wasm.api.WasmInvoker` using the `@WasmTool` annotation.
+This very basic interface wraps te call to the binary, passing the parameters as a byte array and getting the result as a byte array.
+
+[source,java]
+----
+
+    @Inject
+    @WasmTool(value = "greet")
+    org.wildfly.wasm.api.WasmInvoker greet;
+
+    public String greet(String person) {
+        byte[] output = greet.call("greet", person.getBytes(StandardCharsets.UTF_8));
+        return new String(output, StandardCharsets.UTF_8);
+    }
+
+----
+
+A second way is to use the `@WasmToolService` over an interface to create a dynamic proxy wrapping the call to the WASM binary and providing a way to serialize and deserialize your arguments to the byte array required by the invoker.
+
+[source,java]
+----
+@WasmToolService(wasmToolConfigurationName = "pizza", wasmMethodName = "retrievePizzeriaAddresses", argumentSerializer = CityJsonSerializer.class)
```

Comment:
idem: does the `wasmToolConfigurationName` prop correspond to the `wasm-tool` resource's name?

Reply:
Yes, the sentence after the example is ctelling that